### DISCUSSION TITLE
Easy way to print hardware info at the beginning of a session.

### DISCUSSION THREAD

**Discussion by FirefoxMetzger** [pytest-dev/pytest#10744] [Q&A] [answered]
Is there an easy way to print information about the machine's hardware (CPU/RAM/...) at the beginning of a pytest session?

**Background**: I'm working on a numerics library (linalg package for a rendering engine) and I'm trying to get to the bottom of a sporadic failure where a seemingly simple function (norm of a vector) sometimes produces slightly inaccurate results (near machine precision, ie., `1e-15` ballpark). The test works fine locally n all hardware/OS combinations I have access to, but fails about 40% of the time on CI (GH actions). I learned that GH Action runners are not all normalized, so my current guess is that the failure is runner specific. Now I would like to check if this failure correlates with the specific CPU/arch/version and hence my question if pytest can easily add that info to the beginning of a session. (If not, I can also use a dedicated script.)

**Comment by nicoddemus** [comment, marked as answer]
Hi,

You can implement the [pytest_report_header](https://docs.pytest.org/en/stable/reference/reference.html#pytest.hookspec.pytest_report_header) hook in your topmost conftest.py file to add any information to the header.

**Reply by FirefoxMetzger** [reply]
Beautiful! Once again pytest proves to be an awesome library that allows you to do anything you might wish for ⭐
(Also thanks for the fast reply @nicoddemus )

**Comment by FirefoxMetzger** [comment]
For future reference, this is the hook I ended up with:

```python
def pytest_report_header(config, start_path, startdir):
    # report the CPU model to allow detecting platform-specific problems
    if platform.system() == "Windows":
        name = (
            subprocess.check_output(["wmic", "cpu", "get", "name"])
            .decode()
            .strip()
            .split("\n")[1]
        )
        cpu_info = " ".join([name])
    elif platform.system() == "Linux":
        info_string = subprocess.check_output(["lscpu"]).decode()
        for line in info_string.split("\n"):
            if line.startswith("Model name"):
                cpu_info = line[33:]
                break
    else:
        cpu_info = platform.processor()

    return "CPU: " + cpu_info
```

This will add something like the following at the beginning of the logs:
```
CPU: AMD Ryzen 7 5800X 8-Core Processor
# or on a different machine
CPU: Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz
```
